pilot/cmd/pilot-agent/status/server.go

```diff
@@ -508,29 +522,44 @@ func (s *Server) handleStats(w http.ResponseWriter, r *http.Request) {
 		format = expfmt.FmtText
 	}
 
-	if agent, err = scrapeAgentMetrics(); err != nil {
-		log.Errorf("failed scraping agent metrics: %v", err)
-		metrics.AgentScrapeErrors.Increment()
-	}
-
 	w.Header().Set("Content-Type", string(format))
 
 	// Write out the metrics
-	if _, err := w.Write(agent); err != nil {
-		log.Errorf("failed to write agent metrics: %v", err)
+	if err = scrapeAndWriteAgentMetrics(io.Writer(w)); err != nil {
+		log.Errorf("failed scraping and writing agent metrics: %v", err)
 		metrics.AgentScrapeErrors.Increment()
 	}
+
 	if envoy != nil {
-		if _, err := w.Write(envoy); err != nil {
-			log.Errorf("failed to write envoy metrics: %v", err)
-			metrics.EnvoyScrapeErrors.Increment()
+		if format == expfmt.FmtOpenMetrics {
+			envoyStatsBuffer := &bytes.Buffer{}
+			_, err = envoyStatsBuffer.ReadFrom(envoy)
```

Comment:
Isn't there a way to process metrics incrementally instead of reading the whole response in memory?